library/std/src/net/ip/network.rs

```diff
@@ -0,0 +1,392 @@
+#[derive(Copy, PartialEq, Eq, Clone, Hash)]
+#[unstable(feature = "ip_prefix", issue = "86991")]
+pub struct Ipv4AddrPrefix {
+    address: Ipv4Addr,
+    len: u8,
+}
+
+/// Represents an [IPv6 address] prefix, as described in [IETF RFC 4291 section 2.3].
+///
+/// An IPv6 address prefix can be represented by an IPv4 address and a prefix length.
+/// The prefix length specifies how many of the leftmost contiguous bits of the
+/// address comprise the prefix. The prefix then describes the network containing all
+/// addresses that start with the same bits as the prefix.
+///
+/// The size of an `Ipv6AddrPrefix` struct may vary depending on the target operating system.
+///
+/// # Textual representation
+///
+/// `Ipv6AddrPrefix` provides a [`FromStr`](crate::str::FromStr)  implementation.
+/// The textual representation of an address prefix is the representation of an
+/// IPv6 address, followed by a separating `/` and then the prefix length in decimal notation;
+/// for example `2001:db8::/32`.
+///
+/// # Examples
+/// ```
+/// #![feature(ip_prefix)]
+///
+/// use std::net::{Ipv6Addr, Ipv6AddrPrefix};
+///
+/// // Create the address prefix `2001:db8::/32`.
+/// let prefix = Ipv6AddrPrefix::new(Ipv6Addr::new(0x2001, 0xdb8, 0, 0, 0, 0, 0, 0), 32).unwrap();
+///
+/// // The prefix is uniquely defined by a prefix address and prefix length.
+/// assert_eq!(prefix.address(), Ipv6Addr::new(0x2001, 0xdb8, 0, 0, 0, 0, 0, 0));
+/// assert_eq!(prefix.len(), 32);
+///
+/// // The prefix describes the network of all addresses that start
+/// // with the same bits as the prefix.
+/// assert_eq!(prefix.contains(&Ipv6Addr::new(0x2001, 0xdb8, 0, 0, 0, 0, 0, 0)), true);
+/// assert_eq!(prefix.contains(&Ipv6Addr::new(0x2001, 0xdb8, 0, 0, 0, 0, 0, 7)), true);
+/// assert_eq!(prefix.contains(&Ipv6Addr::new(0x2001, 0xdb8, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff)), true);
+/// assert_eq!(prefix.contains(&Ipv6Addr::new(0x2001, 0xdb9, 0, 0, 0, 0, 0, 0)), false);
+/// ```
+///
+/// [IPv6 address]: Ipv6Addr
+/// [IETF RFC 4291 section 2.3]: https://tools.ietf.org/html/rfc4291#section-2.3
+#[derive(Copy, PartialEq, Eq, Clone, Hash)]
+#[unstable(feature = "ip_prefix", issue = "86991")]
+pub struct Ipv6AddrPrefix {
+    address: Ipv6Addr,
+    len: u8,
+}
+
+impl Ipv4AddrPrefix {
+    /// Creates a new IPv4 address prefix from an address and a prefix length.
+    ///
+    /// # Examples
+    /// ```
+    /// #![feature(ip_prefix)]
+    ///
+    /// use std::net::{Ipv4Addr, Ipv4AddrPrefix};
+    ///
+    /// // Create the address prefix `192.0.2.0/24`.
+    /// let prefix = Ipv4AddrPrefix::new(Ipv4Addr::new(192, 0, 2, 0), 24).unwrap();
+    ///
+    /// // Error: Prefix length can not be more than 32 bits.
+    /// Ipv4AddrPrefix::new(Ipv4Addr::new(192, 0, 2, 0), 35).unwrap_err();
+    /// ```
+    #[unstable(feature = "ip_prefix", issue = "86991")]
+    #[inline]
+    pub const fn new(address: Ipv4Addr, len: u32) -> Result<Ipv4AddrPrefix, InvalidPrefixError> {
```

Comment:
Constructor is fallible, unlike the `IpvAddr` and `SocketAddr` constructors, because the prefix length can not be longer than there are bits.